org.eclipse.m2e.jdt/src/org/eclipse/m2e/jdt/internal/ModuleSupport.java

```diff
@@ -319,4 +335,430 @@ public static boolean isMavenJavaProject(IProject project) {
     }
     return false;
   }
+
+  /**
+   * Updates M2E container and JRE container with the jpms arguments provided to the maven-compiler-plugin. The dispatch
+   * between M2E and JRE containers is done using the targeted module. If the JRE container contains the targeted module
+   * then the jpms argument will be attached to it else it will be attached to the M2E container
+   * 
+   * @param request
+   * @param classpath
+   * @param monitor
+   * @param compilerArgs
+   */
+  public static void configureRawClasspath(ProjectConfigurationRequest request, IClasspathDescriptor classpath,
+      IProgressMonitor monitor, List<String> compilerArgs) {
+
+    IMavenProjectFacade facade = request.getMavenProjectFacade();
+
+    IJavaProject javaProject = JavaCore.create(facade.getProject());
+    if(javaProject == null || !javaProject.exists() || classpath == null) {
+      return;
+    }
+
+    JpmsArgs jpmsArgs = JpmsArgs.computeFromArgs(compilerArgs);
+
+    IClasspathEntry m2eEntry = BuildPathManager.getMavenContainerEntry(javaProject);
+    IClasspathEntry jreEntry = BuildPathManager.getJREContainerEntry(javaProject);
+
+    boolean fromScratch = false;
+    if(m2eEntry == null || jreEntry == null) { // .classpath has been deleted or do not exists
+      m2eEntry = findMavenContainerEntry(classpath);
+      jreEntry = findJreContainerEntry(classpath);
+      fromScratch = true;
+    }
+
+    IClasspathEntryDescriptor m2eEntryDescriptor = new ClasspathEntryDescriptor(m2eEntry);
+    IClasspathEntryDescriptor jreEntryDescriptor = new ClasspathEntryDescriptor(jreEntry);
+
+    if(fromScratch) { // if .classpath does not exists
+      try {
+        // we need to set the java project to allow module resolution
+        javaProject.setRawClasspath(classpath.getEntries(), monitor);
+        // we request a classpath resolution to prevent some missing jre modules in the getRootModules call (50/70)
+        javaProject.getResolvedClasspath(false);
+      } catch(JavaModelException ex) {
+        log.error(ex.getMessage(), ex);
+      }
+    }
```

Comment:
I don't believe this is ModuleSupport's responsibility to handle that case

Reply:
AbstractJavaProjectConfigurator is now responsible to update the project prior ModuleSupport execution